pyroclast/common/feature_classifier.py

```diff
@@ -0,0 +1,142 @@
+import abc
+import numpy as np
+import tensorflow as tf
+
+from pyroclast.common.adversarial import fast_gradient_sign_method
+import sonnet as snt
+
+
+class FeatureClassifierMixin(abc.ABC):
+    """Feature Classifier Mixin.
+
+    A mixin that provides base implementations for models that work by
+    first calculating features via a conv stack and then performing
+    classification based off those features. Formalizing this much of
+    the network allows us to perform `usefulness' and `robustness'
+    analysis on the features.
+
+    Definitions of 'usefulness' and 'robustness' from :ref:`Adversarial Examples Are
+    Not Bugs, They Are Features by Ilyas et al.<https://arxiv.org/abs/1905.02175>`
+    """
+
+    @abc.abstractmethod
+    def features(self, x):
+        """Returns a flat tensor of features
+
+        This method calculates the features given some input data x.
+
+        Args:
+           x (tf.Tensor): Input of shape [batch_size, ...data_shape]
+
+        Returns:
+           features (tf.Tensor): Features of each data in the input [batch_size, num_features]
+        """
+        pass
+
+    @abc.abstractmethod
+    def classify_features(self, features):
+        """Calculates output logits given a feature tensor
+
+        Args:
+           features (tf.Tensor): Input of shape [batch_size, num_features]
+
+        Returns:
+           logits (tf.Tensor): Output of shape [batch_size, num_classes]
+        """
+        pass
+
+    def usefulness(self, D):
+        """Finds maximal rho such that f is rho-useful over the dataset D for
```

Comment:
This description is no longer accurate. These docs should be updated.

Reply:
Could you clarify which part isn't accurate anymore? I'm fairly sure nothing essential has changed that requires changing the description.

Reply:
Well, it doesn't say anything about calculating rho for each entry in the cross product of features and classes except in the return type. The initial description makes it seem like we are calculating rho for a single f, c.